### PR TITLE
GH-124478: Cleanup argparse documentation

### PR DESCRIPTION
**Pull request by savannahostrowski**
This PR:
1) Removes duplicative examples that are fairly basic and sufficiently covered in the tutorial
2) Removes the quick reference tables, as removing duplicative examples moves the reference docs for `ArgumentParser` and `add_argument` much higher up the page. I don't think this is terribly necessary anymore.
3) Move the upgrading to optparse guide to a separate page. Since optparse has been deprecated since 3.2, this is probably a minimally viewed part of the main argparse docs, so I think it's fair to have it on a separate page.
4) Consolidates some examples
5) Updates terminology (e.g., sub-command -> subcommand) for consistency.

Please let me know if there are other best practices we consistently use in other documentation areas but lack here. I'm happy to incorporate them here so we can get these documents feeling fresh!

(As a note, I tried to be somewhat conservative here as I do think many of the examples are 1) valuable for new users and 2) too advanced to add to the tutorial in its current form)

* Issue: gh-124478

### CODE REVIEW

File: Doc/library/argparse-optparse.rst [hunk 1]
```diff
@@ -0,0 +1,55 @@
+.. currentmodule:: argparse
+
+.. _upgrading-optparse-code:
```
Comment (JelleZijlstra) [inline Doc/library/argparse-optparse.rst]: I feel this filename makes it look like it's docs for the `argparse.optparse` module. What do you think of moving it into `howto/` instead?

File: Doc/library/argparse-optparse.rst [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+==========================
+
+Originally, the :mod:`argparse` module had attempted to maintain compatibility
+with :mod:`optparse`.  However, :mod:`optparse` was difficult to extend
+transparently, particularly with the changes required to support the new
+``nargs=`` specifiers and better usage messages.  When most everything in
+:mod:`optparse` had either been copy-pasted over or monkey-patched, it no
+longer seemed practical to try to maintain the backwards compatibility.
+
+The :mod:`argparse` module improves on the standard library :mod:`optparse`
```
Comment (JelleZijlstra) [inline Doc/library/argparse-optparse.rst]:
```suggestion
The :mod:`argparse` module improves on the :mod:`optparse`
```
They're both in the standard library, I suspect this line comes from when argparse was a third-party module.

File: Doc/library/argparse-optparse.rst [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+.. _upgrading-optparse-code:
+    
+==========================
+Upgrading optparse code
+==========================
+
+Originally, the :mod:`argparse` module had attempted to maintain compatibility
+with :mod:`optparse`.  However, :mod:`optparse` was difficult to extend
+transparently, particularly with the changes required to support the new
```
Comment (JelleZijlstra) [inline Doc/library/argparse-optparse.rst]:
```suggestion
transparently, particularly with the changes required to support
```
Not remotely new any more

File: Doc/library/argparse-optparse.rst (hunk 1 above)
Comment (savannahostrowski) [reply Doc/library/argparse-optparse.rst]: Yeah - that's fair. Let me move that.

File: Doc/library/argparse.rst [hunk 4]
```diff
@@ -892,14 +663,7 @@ how the command-line arguments should be handled. The supplied actions are:
 * ``'store_true'`` and ``'store_false'`` - These are special cases of
   ``'store_const'`` used for storing the values ``True`` and ``False``
   respectively.  In addition, they create default values of ``False`` and
-  ``True`` respectively.  For example::
```
Comment (JelleZijlstra) [inline Doc/library/argparse.rst:895]: I feel this example is probably helpful to keep. Not everyone is going to read the tutorial, and examples are extremely useful for people just glancing across the docs page to find the thing they need.

File: Doc/library/argparse.rst [hunk 5]
```diff
@@ -1260,28 +1024,11 @@ Some command-line arguments should be selected from a restricted set of values.
 These can be handled by passing a sequence object as the *choices* keyword
 argument to :meth:`~ArgumentParser.add_argument`.  When the command line is
 parsed, argument values will be checked, and an error message will be displayed
-if the argument was not one of the acceptable values::
```
Comment (JelleZijlstra) [inline Doc/library/argparse.rst]: Similarly, I'd keep at least one example of `choices`, probably the first one.

File: Doc/library/argparse.rst [hunk 6]
```diff
@@ -2400,3 +2080,15 @@ Exceptions
 .. exception:: ArgumentTypeError
 
    Raised when something goes wrong converting a command line string to a type.
+
+.. We use the "rubric" directive here to avoid creating
```
Comment (savannahostrowski) [inline Doc/library/argparse.rst]: I copied this from the asyncio docs. There might be a better way to do this?

File: Doc/library/argparse.rst [hunk 7]
```diff
@@ -1806,7 +1538,7 @@ Sub-commands
    Description of parameters:
 
    * title - title for the sub-parser group in help output; by default
-     "subcommands" if description is provided, otherwise uses title for
+     "sub-commands" if description is provided, otherwise uses title for
```
Comment (JelleZijlstra) [inline Doc/library/argparse.rst]: The default is actually "subcommands" at runtime, not "sub-commands" (line 1826 of argparse.py at the moment). We should probably standardize on that spelling in the docs too.

Comment (savannahostrowski) [reply Doc/library/argparse.rst]: Ah, right. That makes sense. I'll err on the side of "subcommands"

File: Doc/library/argparse.rst (hunk 6 above)
Comment (JelleZijlstra) [reply Doc/library/argparse.rst]: On asyncio the "Reference" section points to the reference docs for individual modules within the asyncio package, but that's not the case here, so I don't think the name "Reference" works here. Maybe use "Guides and Tutorials" as a normal heading (included in the TOC) and remove the `:caption:`.

**Comment by JelleZijlstra** [comment]
I looked at the preview https://cpython-previews--124877.org.readthedocs.build/en/124877/library/argparse.html#module-argparse and skimmed the entire doc. Below are some things I noticed. None are new to this PR, so feel free to ignore them and I can file a separate PR with those cleanups.

- It's distracting that the first few paragraphs link ``` :mod:`argparse` ``` 4 times, and it just links back to the top of the page. Would recommend suppressing those links with ``` :mod:`!argparse` ```.
- RawTextHelpFormatter description says "new lines"; I think it's normally spelled "newlines".
- "[ArgumentParser](https://cpython-previews--124877.org.readthedocs.build/en/124877/library/argparse.html#argparse.ArgumentParser) uses [filesystem encoding and error handler](https://cpython-previews--124877.org.readthedocs.build/en/124877/glossary.html#term-filesystem-encoding-and-error-handler) to read the file containing arguments." this sentence reads oddly, maybe needs a "the"
- "the action default will not over write it" -> I think "overwrite" is one word
- I might remove the `parser.add_argument('source_file', type=open)` example. It doesn't make it easy to properly close the file, and as the text says later, FileNotFoundError would not be handled nicely.
- `FileType` is formatted inconsistently in the paragraph that starts `Even [FileType](https://cpython-previews--124877.org.readthedocs.build/en/124877/library/argparse.html#argparse.FileType) has its limitations for use with the type keyword`.
- The text after "class argparse.Action" should be indented by another level.
- The docs for `parse_known_intermixed_args` are not properly indented.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by savannahostrowski** [comment]
Thanks so much for such a thorough review, @JelleZijlstra. I think I can incorporate your feedback in this PR. I'll revisit this tomorrow evening.

File: Doc/library/argparse.rst [hunk 8]
```diff
@@ -1825,20 +1573,20 @@ Sub-commands
    * dest_ - name of the attribute under which sub-command name will be
      stored; by default ``None`` and no value is stored
 
-   * required_ - Whether or not a subcommand must be provided, by default
+   * required_ - Whether or not a sub-command must be provided, by default
```
Comment (tomasr8) [inline Doc/library/argparse.rst]:
```suggestion
   * required_ - Whether or not a subcommand must be provided, by default
```

File: Doc/howto/argparse-optparse.rst [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+transparently, particularly with the changes required to support
+``nargs=`` specifiers and better usage messages.  When most everything in
+:mod:`optparse` had either been copy-pasted over or monkey-patched, it no
+longer seemed practical to try to maintain the backwards compatibility.
+
+The :mod:`argparse` module improves on the :mod:`optparse`
+module in a number of ways including:
+
+* Handling positional arguments.
+* Supporting sub-commands.
```
Comment (tomasr8) [inline Doc/howto/argparse-optparse.rst]: couple more ;)
```suggestion
* Supporting subcommands.
```

**Review by willingc — APPROVED** [review]
Lovely work @savannahostrowski. Thank you!

**Comment by savannahostrowski** [comment]
@JelleZijlstra, I applied all of your suggestions. I also found formatting inconsistencies in taking another pass related to indentation and style.

File: Doc/library/argparse.rst [hunk 10]
```diff
@@ -1288,17 +1289,17 @@ this API may be passed as the ``action`` parameter to
    Action instances should be callable, so subclasses must override the
    ``__call__`` method, which should accept four parameters:
 
-   * ``parser`` - The ArgumentParser object which contains this action.
+   * parser - The ArgumentParser object which contains this action.
```
Comment (JelleZijlstra) [inline Doc/library/argparse.rst]:
```suggestion
   * *parser* - The ArgumentParser object which contains this action.
```
Parameters are conventionally formatted this way (same for the next few ones)

Comment (savannahostrowski) [reply Doc/library/argparse.rst]: Got it - this doesn't apply for parameters that are also jump targets right? Or is there a way to wrap something like `prog_` in * ?

Comment (JelleZijlstra) [reply Doc/library/argparse.rst]: I think I remember someone finding a way to do that but it's very weird and probably not worth it.

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**